src/range.rs

```diff
@@ -182,6 +182,69 @@ pub fn take_while1<I, F>(f: F) -> TakeWhile1<I, F>
     TakeWhile1(f, PhantomData)
 }
 
+pub struct TakeUntil<I>(I::Range) where I: RangeStream;
+impl<I> Parser for TakeUntil<I>
+    where I: RangeStream,
+          I::Range: PartialEq + ::primitives::Range
+{
+    type Input = I;
+    type Output = I::Range;
+
+    #[inline]
+    fn parse_lazy(&mut self, mut input: Self::Input) -> ConsumedResult<Self::Output, Self::Input> {
+        use ::primitives::Range;
+
+        let len = self.0.len();
+        let mut to_consume = 0;
+        let mut look_ahead_input = input.clone();
+    
+        loop {
+            match look_ahead_input.clone().uncons_range(len) {
+                Ok(xs) => {
+                    if xs == self.0 {
+                        break;
+                    } else {
+                        to_consume += 1;
+                        if look_ahead_input.uncons().is_err() {
+                            unreachable!();
+                        }
+                    }
+                },
+                Err(_) => return EmptyErr(ParseError::empty(look_ahead_input.position())),
+            };
+        }
+
+        match input.uncons_range(to_consume) {
+            Ok(x) => ConsumedOk((x, input)),
+            Err(_) => unreachable!(),
+        }
+    }
+}
+
+
+/// Zero-copy parser which reads a range of 0 or more tokens until `r` is found.
+///
+/// The range `r` will not be consumed. If `r` is not found, the parser will
+/// return an error.
+///
+/// ```
+/// # extern crate combine;
+/// # use combine::range::{range, take_until};
+/// # use combine::*;
+/// # fn main() {
+/// let mut parser = take_until("\r\n");
+/// let result = parser.parse("To: user@example.com\r\n");
+/// assert_eq!(result, Ok(("To: user@example.com", "\r\n")));
+/// let result = parser.parse("Hello, world\n");
+/// assert!(result.is_err());
+/// # }
+/// ```
+#[inline(always)]
+pub fn take_until<I>(r: I::Range) -> TakeUntil<I> where I: RangeStream
```

Comment:
I'd probably call it `take_until_range` to leave the possibility open for `take_while_range` and `take_until` (which would be equivalent to `take_while(|t| !predicate(t))`)